### ISSUE TITLE
Add user folder support for subfolders

### ISSUE THREAD

**Issue by quisquous** [quisquous/cactbot#860]
It's cumbersome to stick all the logic into a single file in the user folder.  It'd be easier if folks could share whole files that could be dropped into

One possibility here is to not just load the javascript file named the same as the overlay, but also the subfolder (and anything underneath it).  In other words, `user/raidboss.js` is loaded, but also any javascript/css files under `user/raidboss/*`.  Maybe we'd want to ignore `.git` directories (or all dot directories?) in case folks put these under version control.

**Comment by Bluefissure** [comment]
Yes! this is exactly what I'm about to discuss.
Subfolder in user makes it much easier for users to manage their own triggers of raidboss.

**Comment by quisquous** [comment]
The one downside here is that you can't have a "remote" user directory, because you can't enumerate subfolders remotely.  Maybe that's fine, but I'm not sure how to document that.

**Comment by Bluefissure** [comment]
> The one downside here is that you can't have a "remote" user directory in this same way, because you can't enumerate subfolders remotely. Maybe that's fine, but I'm not sure how to document that.
> 
> One thing that might be useful to is to be able to load trigger files remotely, such that somebody could have a github repo with cactbot triggers for their raid, and just have everybody (once) include a file that says "load this github file too". The downside of this is that you can't do this for "local" user files directly in js because of browser security restrictions. But we could add a new handler that says "please fetch this js file and hand it to me".
> 
> Some more things to think about. :)

The remote site can set `autoindex on` to host user files, but that requires a server, not Github pages.
Moreover, a metafile seems good, indicating all the paths of user files.

Personally I don't think many players will use remote raidboss, since it will load tons of data files?

**Comment by quisquous** [comment]
A surprising number of people use remote raidboss, even though it loads tons of data files.

I think my thought here for how to handle this is that maybe it's fine to disallow remote user directories, and make people keep it local.  I think there could be a separate option to load a remote user javascript file by name as well, which would allow previous behavior to keep working.

**Comment by valarnin** [comment]
I was thinking about tackling this, but given the discussions we've had on other issues here and via Discord, I'm not sure if this is still the path you want to go with this?

**Comment by quisquous** [comment]
I think there's a few questions to sort out before starting, that I'm not sure I have good answers for.  (I did talk about remote cactbot directories above, but I looked through the code and there is no remote user directory support.  Even if it attempts to load a directory adjacent to a remote url it's only operating on local paths.)

1) Improved Security: ngld has not implemented this in their merged OverlayPlugin+cactbot because it's a security hole (load files from an arbitrary folder on disk), and so it might be worth soliciting their opinions about what the shape of this could look like.

2) Types of files in the subdirectory: right now `user/raidboss.js` is a file that is manipulating `Options`.  Should files in the subdirectory be the same sort of thing? Or should they look more like trigger files so folks can more easily copy and paste from cactbot triggers? Or is it confusing to have different types of files, and better to keep it consistent?

3) What files to load in the subdirectory: I think just any .js file underneath, ignoring .git subfolders? But maybe there's more nuances here.  Probably it would be good to have a defined order to load them in too for consistency.

**Comment by valarnin** [comment]
Pinging @ngld  for feedback.

What I propose is:

1. Plugin-side the existing logic should be kept in place (load `<user config path>/<overlay>.js`, `<user config path>/<overlay>.css`).
2. In addition to these, add `<user config path>/<overlay>/**`.
3. Before reading any files/sending back to client, exclude any file or folder that starts with a `.`, e.g. `.git`.

It would then be up to the overlay to decide what to do with each file. For instance, raidboss might implement the following logic:
1. Check `e.detail.localUserFiles` for `raidboss.js` key, process it as normal and remove it from the files array
2. Repeat for `raidboss.css` key
3. For all remaining keys in `e.detail.localUserFiles`, treat any key that ends with `.js` as a trigger file and any key that ends with `.txt` as a timeline file, and pass them through as appropriate.

Upside is, it's not a breaking change. Existing setups will continue to work as they are. Downside is that user config will continue to not work from a URL.

As for security, I'm not sure exactly how that would be handled.

- Make sure `<overlay>` parameter that's passed to the plugin is a valid folder name (e.g. doesn't contain any special characters).
- Require that `<overlay>.js` exists in the target user config directory before doing anything else, and only allow known overlay names?
- Maybe filter it to only allow `%APPDATA%\Advanced Combat Tracker\**` or `%USERPROFILE%\Documents\My Games\Advanced Combat Tracker\**`?
- Additionally, I'm not sure exactly how it would be done, but make sure the read is done with the user's permissions instead of the run-as-admin ACT process's permissions?
- It's important to note that the user config path is currently set within ACT itself, it's not passed in to the handler. This limits the potential security issues significantly, since the user themselves would have to go and configure the user config path to be somewhere that would cause problems.

**Comment by quisquous** [comment]
Given that @ngld is working on integrating cactbot c# code into OverlayPlugin, at the *very* least it would be worth coordinating to make sure anything done here doesn't make life harder there, and possibly worth waiting on this until he's done releasing and doing that work there (depending on timing?).  Mostly I'd like to have a conversation before landing anything.

> What I propose is:
> 
> 1. Plugin-side the existing logic should be kept in place (load `<user config path>/<overlay>.js`, `<user config path>/<overlay>.css`).
> 2. In addition to these, add `<user config path>/<overlay>/**`.
> 3. Before reading any files/sending back to client, exclude any file or folder that starts with a `.`, e.g. `.git`.

We could probably filter by `<overlay>` in C# as well, for step 1 and step 2?

> Upside is, it's not a breaking change. Existing setups will continue to work as they are. Downside is that user config will continue to not work from a URL.

I think it's fine to not make this work, as I don't think this actually works right now.

> As for security, I'm not sure exactly how that would be handled.

I'm going to file a separate issue on this front with some comments from ngld on a discussion we had in #1767.

**Comment by quisquous** [comment]
> It would then be up to the overlay to decide what to do with each file. For instance, raidboss might implement the following logic:
> 
> 1. Check `e.detail.localUserFiles` for `raidboss.js` key, process it as normal and remove it from the files array
> 2. Repeat for `raidboss.css` key
> 3. For all remaining keys in `e.detail.localUserFiles`, treat any key that ends with `.js` as a trigger file and any key that ends with `.txt` as a timeline file, and pass them through as appropriate.

(3) here is also a bit up in the air to me.  I think there's a difference between trigger files, timeline files, and user js files.  In the sense that a trigger file is eval'd, but a user js file is just included (and can touch a lot more).

I think it probably it would be better to treat js files in subdirectories as just additional user js files (vs trigger files?), because it leaves a lot more options if somebody wants to make some invasive changes (nicknames, other code changes?) and then distribute some folder to their raid group.  I'm a little torn though because this is a pretty big point of confusion for a lot of people to have user js files be so different than trigger js files.

**Comment by quisquous** [comment]
The primary use case of user folders is for users to copy and paste trigger files from cactbot in whole into the user folder, and then edit them down there.

Some additional issues here are that I think it would be good to have trigger files be modules.  Right now, trigger files are eval'd, and the final value (which is the trigger set, vs any helper functions and data that came earlier) are the trigger set.  I think having to `eval` them is a bit suspect.  In a webpack world, we can import them all instead of eval-ing.  This'll let us (more easily) minify the source etc etc.  It also just makes the dependencies clearer.  If some trigger file wants to import something else, it's currently not possible.

If we want to change the trigger file format, then this creates an issue of backcompat.  If user folders ship before Webpack, then users will copy non-module trigger files into user folders and we'll need to support those or make (yet another) breaking change.  I am not particularly keen on making this breaking change.  It's probably rough enough for users to switch off of modifying files directly as it is.

So, given all of these constraints, I see a couple options:
1) hold off on adding user folders until Webpack lands and we can convert the trigger files to modules properly
2) convert the trigger files now *as if* they were modules, but then do some hacky regex work prior to eval to remove imports and exports so that they continue to be eval'd until some future point when Webpack lands and they can be treated as modules properly.

(2) seems incredibly hacky to me but also will make user triggers a lot easier to make in the short term before Webpack.  I'm a little bit torn.

**Comment by quisquous** [comment]
I'm actually having some second thoughts about how files in the user folder could also be modules.  cactbot itself would see either string contents or a file location on disk, and neither of those can use imports.  Even if somehow they could use import statements, what would the path for those imports be?  I think for all of these reasons that the hypothetical use case of "copy and paste trigger files into user folder" is not possible.

I still think it's worth converting triggers to be modules themselves and having explicit imports.

So, I think then the only remaining option is that files in user subfolders behave exactly like files directly in the user folder, just with some specified eval ordering.  (I would suggest alphabetical order, except subfolders before root folders, so that `user/raidboss.js` can have a chance to override `user/raidboss/myraidgroup/some_triggers.js`.)
